implementations/rust/ockam/ockam_node/src/runtime.rs

```diff
@@ -14,3 +14,106 @@ pub fn take() -> Runtime {
         .take()
         .expect("Runtime was consumed")
 }
+
+#[cfg(feature = "watchdog")]
+pub(crate) mod watchdog {
+    use std::collections::HashMap;
+    use std::thread::ThreadId;
+    use std::time::Duration;
+    use tokio::runtime::Runtime;
+
+    struct ThreadInfo {
+        poll_count: u64,
+        was_parked: bool,
+        thread_id: ThreadId,
+    }
+
+    /// Detects if there is any thread stuck on a task and write a warning log about it.
+    /// Print its stacktrace if possible.
+    pub(crate) struct TokioRuntimeWatchdog {
+        thread_map: HashMap<usize, ThreadInfo>,
+    }
+
+    impl TokioRuntimeWatchdog {
+        pub(crate) fn new() -> Self {
+            Self {
+                thread_map: HashMap::new(),
+            }
+        }
+
+        pub fn start_watchdog_loop(self, runtime: &Runtime) {
+            let metrics = runtime.metrics();
+            // to avoid spawning the watchdog within tokio runtime
+            std::thread::spawn(move || {
+                self.watchdog_loop(metrics);
+            });
+        }
+
+        fn watchdog_loop(mut self, metrics: tokio::runtime::RuntimeMetrics) {
+            info!("Starting tokio runtime watchdog");
+            const WATCHDOG_INTERVAL: Duration = Duration::from_millis(50);
+            let watchdog_interval = std::env::var("OCKAM_WATCHDOG_INTERVAL")
+                .map(|s| Duration::from_millis(s.parse().unwrap()))
+                .unwrap_or(WATCHDOG_INTERVAL);
+
+            loop {
+                // create any missing thread info
+                for worker in 0..metrics.num_workers() {
```

Comment:
Took me a while to understand that it's tokio internal metrics and those are not Ockam workers. But not sure how we could articulate that better here

Reply:
:) I used `tokio_metrics` and `tokio_worker`